### ISSUE TITLE
ERROR ON runRndcActions CLIENT DNS

### ISSUE THREAD

**Issue by nikpael** [WillyXJ/facileManager#636]
Please prefix your issue title with one of the following: [BUG]

Replace everything between stars with current version of your facileManager and module installations:  
fM Version : **4.7.o**
<Specify Module Name> Version : **6.2.0**

**In raising this issue, I confirm the following (please check boxes, eg [X]):**

- [x] I have read and understood the [contributors guide](https://github.com/WillyXJ/facileManager/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that the bug-fix I am reporting can be replicated, or that the feature I am suggesting isn't already present.
- [x] I have checked that the issue I'm posting isn't already reported.
- [x] I have checked that the issue I'm posting isn't already solved and no duplicates exist in [closed issues](https://github.com/WillyXJ/facileManager/issues?q=is%3Aissue+is%3Aclosed) and [opened issues](https://github.com/WillyXJ/facileManager/issues)
- [x] I have checked the pull requests tab for existing solutions/implementations to my issue/suggestion.

---

**(BUG | ISSUE) Expected Behavior:**
reload and update zone and server config

**(BUG | ISSUE) Actual Behavior:**
error on facilemanager:
ns1.idealogic.it
Configuration Build Results
[ns1.xx.it] rndc: connection to remote host closed
[ns1.xx.it] This may indicate that
[ns1.xx.it] * the remote server is using an older version of the command protocol,
[ns1.xx.it] * this host is not authorized to connect,
[ns1.xx.it] * the clocks are not synchronized,
[ns1.xx.it] * the key signing algorithm is incorrect, or
[ns1.xx.it] * the key is invalid.
[ns1.xx.it] rndc: connection to remote host closed
[ns1.xx.it] This may indicate that
[ns1.xx.it] * the remote server is using an older version of the command protocol,
[ns1.xx.it] * this host is not authorized to connect,
[ns1.xx.it] * the clocks are not synchronized,
[ns1.xx.it] * the key signing algorithm is incorrect, or
[ns1.xx.it] * the key is invalid.

on bind9 log:  invalid command from 127.0.0.1#55189: bad auth

**(BUG | ISSUE) Steps to reproduce:**
reload zone or building server config

{replace this line with your issue content}
I compared the current version with the previous one in /usr/local/facileManager/fmDNS/functions.php and the error seems to be in the runRndcActions() function. if I put global $server_config_file back in it; everything works

**Comment by WillyXJ** [comment]
So far I don't see how it would work with the global statement since the `$server_config_file` variable gets passed to the function correctly. I'm currently unable to reproduce it. My test environment did have some bad keys from early tests that prevented rndc from controlling bind like the error you got. I ended up regenerating my keys with `rndc-confgen -a`, restarted named, and all was well.

So, did your keys change?

**Comment by nikpael** [comment]
> So far I don't see how it would work with the global statement since the `$server_config_file` variable gets passed to the function correctly. I'm currently unable to reproduce it. My test environment did have some bad keys from early tests that prevented rndc from controlling bind like the error you got. I ended up regenerating my keys with `rndc-confgen -a`, restarted named, and all was well.
> 
> So, did your keys change?

my keys should not have changed. if I run rndc from the command line on the name server it works. I also tried downgrading both the server and the facilemanager client without uninstalling bind and it works again.

**Comment by WillyXJ** [comment]
What version did you downgrade to?

**Comment by nikpael** [comment]
> What version did you downgrade to?

4.6.2 server and 6.0.6 client. i think they are the previous

**Comment by WillyXJ** [comment]
> Adding back the `global $server_config_file;` in the first line of the function solves the problem. Can I ask why you added this `$server_config_file = null` statement as function's argument? Please be kind... I barely can read PHP! :$

$server_config_file was moved to be a function parameter to eliminate the need for referencing a global variable. runRndcActions() is referenced twice and both occurrences are passing a value for $server_config_file so it makes no sense to me why adding back the global statement in the function would fix it.

I'm also still unable to reproduce the problem in my lab. Can somebody please provide me with specific reproducible steps?  I have a feeling there's a server and/or key configuration that I'm not applying that would make this reproducible.

**Comment by andrewhotlab** [comment]
After a couple of additional tests, I found that the problem arises when the file `named.conf.keys` exists. I defined some TSIG keys in my environment, and I suppose it's the same for @nikpael.

As I wrote, I'm not a developer so the code is barely understandable by me, but I think that this line adds the parameter "-k [confdir]/named.conf.key" to the `rndc` command:
```
$rndc .= (file_exists(dirname($server_config_file) . '/named.conf.keys')) ? sprintf(' -k %s/named.conf.keys', dirname($server_config_file)) : null;
```
But the `rndc` key is saved by default in the `rndc.key` file by deafult. In fact, if I run `rndc -k [confdir]/named.conf.keys [action]` from command line on the same host I receive this output, which is the same returned by the web GUI:

```
rndc: connection to remote host closed.
* This may indicate that the
* remote server is using an older
* version of the command protocol,
* this host is not authorized to connect,
* the clocks are not synchronized,
* the key signing algorithm is incorrect
* or the key is invalid.
```
Maybe the reason why adding the `global $server_config_file;` apparently solves the problem is that the above line makes the $rndc addition null, and so no "-k" parameter is added to the `rndc` command.

I do not know if the same behavior  exists on other platforms, but on FreeBSD `rndc` reads its key in the file `/usr/local/etc/named/rndc.key` by default, thus there is no need to add the "-k" parameter, or it might make sense to change that line in such a way:
```
$rndc .= (file_exists(dirname($server_config_file) . '/rndc.key')) ? sprintf(' -k %s/rndc.key', dirname($server_config_file)) : null;
```

**Comment by nikpael** [comment]
yes i also have a key in `named.conf.keys` could this be the problem from the code you sent.

**Comment by WillyXJ** [comment]
I think this is a result of #389 that was implemented in v6.0.0. I will work on adding an option allowing you to use named.conf.keys with rndc (some people may want it while others not).  In the meantime, a temporary fix can be to comment out https://github.com/WillyXJ/facileManager/blob/d2df1dd843a2f23343bd9b3c0bc7c4cca60caefc/client/facileManager/fmDNS/functions.php#L548.

**Comment by nikpael** [comment]
thanks for the work around. in reality I save in named.conf.key a key (key1) that I then use to update with nsupdate a zone (allow-update and value key1). it almost seems that since there is the named.conf.key file, whatever the content, the key for the rndc command is taken

**Comment by WillyXJ** [comment]
This is now fixed in [fmDNS 7.0-beta1](https://www.facilemanager.com/modules/fmdns/?t=3) and later.
